SU2_CFD/src/solvers/CNEMOEulerSolver.cpp

```diff
@@ -956,33 +981,140 @@ void CNEMOEulerSolver::Upwind_Residual(CGeometry *geometry, CSolver **solver_con
     }
   }
 
-  delete [] Conserved_i;
-  delete [] Conserved_j;
-  delete [] Primitive_i;
-  delete [] Primitive_j;
-  delete [] dPdU_i;
-  delete [] dPdU_j;
-  delete [] dTdU_i;
-  delete [] dTdU_j;
-  delete [] dTvedU_i;
-  delete [] dTvedU_j;
-  delete [] Eve_i;
-  delete [] Eve_j;
-  delete [] Cvve_i;
-  delete [] Cvve_j;
+  /*--- Warning message about non-physical reconstructions. ---*/
+  if ((iMesh == MESH_0) && (config->GetComm_Level() == COMM_FULL)) {
+    /*--- Add counter results for all threads. ---*/
+    SU2_OMP_ATOMIC
+    ErrorCounter += counter_local;
+    SU2_OMP_BARRIER
+
+    /*--- Add counter results for all ranks. ---*/
+    SU2_OMP_MASTER
+    {
+      counter_local = ErrorCounter;
+      SU2_MPI::Reduce(&counter_local, &ErrorCounter, 1, MPI_UNSIGNED_LONG, MPI_SUM, MASTER_NODE, MPI_COMM_WORLD);
+      config->SetNonphysical_Reconstr(ErrorCounter);
+    }
+    SU2_OMP_BARRIER
+  }
+
+}
+
+su2double CNEMOEulerSolver::ComputeConsistentExtrapolation(CNEMOGas *fluidmodel, unsigned short nSpecies, su2double *V,
+                                                           su2double* dPdU, su2double* dTdU, su2double* dTvedU,
+                                                           su2double* val_eves, su2double *val_Cvves) {
+
+  //NOTE: TODO - this doesnt compute Cvves/ dPdU,etc.yet
+  su2double val_gamma;
+  vector<su2double> rhos;
+
+  /*--- Rename index information ---*/
+  unsigned short T_INDEX   = nSpecies;
+  unsigned short TVE_INDEX = nSpecies+1;
+
+  /*--- Rename density vector ---*/
+  rhos.resize(nSpecies,0.0);
+  for (unsigned short iSpecies=0; iSpecies < nSpecies; iSpecies++ ){
+    rhos[iSpecies] = V[iSpecies];
+  }
+
+  /*--- Set new fluid state ---*/
+  fluidmodel->SetTDStateRhosTTv(rhos, V[T_INDEX], V[TVE_INDEX]);
+
+  /*---Compute the secondary values ---*/
+  auto it = val_eves;
+  auto& ref = fluidmodel->ComputeSpeciesEve(V[TVE_INDEX]);
+  for (auto v : ref) {
+    *it = v;  ++it;
+  }
+  val_eves  = it;
```

Comment:
I would put a comment here explaining you are copy from the vector to returned by the fluid model to val_eves.
Also this last line `val_eves  = it;` can be deleted.

Reply:
@pcarruscag if we delete  `val_eves = it;` how do we have what we want in the vector val_eves that we are trying to compute?

also, in general, can you explain this procedure? lines 1027-1029 - I don't understand this advanced c++ thing and probably it would be better to know 100% what we're doing here, thanks! :)

Reply:
equivalent to:
```
for (int i = 0; i < ref.size(); ++i)
  val_eves[i] = ref[i];
```
but more generic in case you change the types involved.

val_eves is a pointer, i.e. a memory location, if you write val_eves = something you are changing _where_ it points to, instead of _what_ it points to (the loop is what changes the _what_).
It would be less confusing to use a consistent type across the implementation, e.g. std::vector, instead of mixing it with c-style arrays (bare pointers), because a vector can be downgraded to pointer when it needs to be passed to some legacy interface, but to go from pointer to vector you need to copy.
Even better is to write algorithms that do not depend on a specific data structures, but rather on an interface ([], *, ++ are interfaces) that is why the algorithms in the standard template library are written in terms of iterators (but as the name implies that requires templates).

Reply:
ok I see there's still a lot to it generically speaking, but I understand the code we have now, thanks

Reply:
Done